File: Problem2.cpp
//Question: 2.Knapsack problem

// Time Complexity : O(m*n)
// Space Complexity : O(m*n)
// Any problem you faced while coding this : no.
// Did this code successfully run on Leetcode : yes

// Your code here along with comments explaining your approach
// Idea here is to calculate the max gain in each scenario as maximum between the previously obtained 
// gain and gain from one step back. 

#include<vector>
#include <iostream>

int knapSack(int cap, std::vector<int>& weights, std::vector<int>& profit) {
    std::vector<std::vector<int>> result(weights.size()+1, std::vector<int>(cap+1, 0)); 
    for(int i = 1; i <= weights.size(); i++)
    {
        for(int j = 1; j <= cap; j++)
        {
            if(weights[i] > j)
            {
                result[i][j] = result[i-1][j];
            }
            else
            {
                result[i][j] = std::max(result[i-1][j], (profit[i] + result[i-1][j-weights[i-1]]));
            }
        }
    }
    return result[weights.size()][cap];
}


// Driver code 
int main() 
{ 
    std::vector<int> profit = { 60, 100, 120 }; 
    std::vector<int> weight = { 10, 20, 30 }; 
    int capacity = 50; 
    int n = profit.size();
    if(n == 0)
    {
        return 0;
    }

    if(n == 1)
    {
        return weight[0];
    }
    std::cout << "Max gain is: " << knapSack(capacity, weight, profit) << std::endl; 
    return 0; 
}

File: Problem2.js
// Knapsack

const knapSack = (maxCapacity, valueList, weightList, size) => {
    let K = new Array(size + 1);

    for (let i = 0; i <= size; i++) {
        K[i] = new Array(maxCapacity + 1);
        for (let w = 0; w <= maxCapacity; w++) {
            // base
            if (i === 0 || w === 0) {
                K[i][w] = 0;
            }
            // if selected weight <= capacity weight
            else if (weightList[i-1] <= w) {
                // k[i][w] = Math.max(include, exclude)
                K[i][w] = Math.max(valueList[i-1] + K[i-1][w - weightList[i-1]] , K[i-1][w]);
            } 
            // if it exceeds
            else {
                K[i][w] = K[i-1][w];
            }
        }
    }
    return K[size][maxCapacity];
}

let valueList = [60, 100, 120];
let weightList = [10, 20, 30];
let maxCapacity = 50;
let size = valueList.length;
console.log(knapSack(maxCapacity, valueList, weightList, size));